File: BlotoutAnalytics/BlotoutAnalyticsAll/Network/api/BOBaseAPI.m
//
//  BOBaseAPI.m
//  BlotoutAnalytics
//
//  Copyright © 2019 Blotout. All rights reserved.
//

#import "BOBaseAPI.h"
#import "BONetworkManager.h"
#import <BlotoutFoundation/BOFLogs.h>
#import "BOASDKManifestController.h"
#import "BlotoutAnalytics_Internal.h"

@implementation BOBaseAPI

-(NSDictionary*)getJsonData:(NSData*)data {
  NSDictionary *dict = nil;
  @try {
    if (data != nil) {
      dict = [NSJSONSerialization JSONObjectWithData:data options:NSJSONReadingMutableLeaves error:nil];
    }
  } @catch (NSException *exception) {
    BOFLogDebug(@"%@:%@", BOA_DEBUG, exception);
  }
  
  return dict;
}

-(NSString*)getBaseServerUrl {
  return [BlotoutAnalytics sharedInstance].endPointUrl;
}

-(NSString*)resolveAPIEndPoint:(BOUrlEndPoint)endPoint {
  NSString *url;
  switch (endPoint) {
      case BOUrlEndPointEventPublish:{
      url = [NSString stringWithFormat:@"%@/%@", [self getBaseServerUrl],BO_SDK_REST_API_EVENTS_PUSH_PATH];
          break;
      }
    case BOUrlEndPointManifestPull:
      url = [NSString stringWithFormat:@"%@/%@", [self getBaseServerUrl], BO_SDK_REST_API_MANIFEST_PULL_PATH];
  }
  
  return [NSString stringWithFormat:@"%@?token=%@", url, [BlotoutAnalytics sharedInstance].token];
}

-(NSDictionary*)prepareRequestHeaders {
  return @{
      BO_ACCEPT: @"application/json",
      BO_CONTENT_TYPE: @"application/json"
  };
}

/* This Method check for null value in response data and replace it with empty string, Temp fix for Manifest Response */
-(NSData*)checkForNullValue:(NSData*)data {
  NSString *responseString = [[NSString alloc] initWithData:data encoding:NSUTF8StringEncoding];
  
  if ([responseString containsString:@"null,"]) {
      responseString = [responseString stringByReplacingOccurrencesOfString:@"null," withString:@""];
  }
  
  return [responseString dataUsingEncoding:NSUTF8StringEncoding];
}
@end

File: BlotoutAnalytics/Network/api/BOBaseAPI.swift
//
//  BOBaseAPI.swift
//  BlotoutAnalytics
//
//  Created by Poonam Tiwari on 19/04/22.
//

import Foundation


public enum BOUrlEndPoint : Int {
     case eventPublish = 0
     case manifestPull
 }

let EPAPostAPI = "POST"
let EPAGetAPI = "GET"
let EPAContentApplicationJson = "application/json"

class BOBaseAPI:NSObject {
    
    func getJsonData(_ data: Data?) -> [AnyHashable : Any]? {
        var dict: [AnyHashable : Any]? = nil
        if let data = data {
            do {
                dict = try JSONSerialization.jsonObject(with: data, options: .mutableLeaves) as? [AnyHashable : Any]
            } catch {
                BOFLogDebug(frmt: "%@:%@", args: BOA_DEBUG, error.localizedDescription)
            }
        }
        return dict
    }
    
    func getBaseServerUrl() -> String {
        return BlotoutAnalytics.sharedInstance.endPointUrl ?? ""
    }
    
    public func resolveAPIEndPoint(_ endPoint: BOUrlEndPoint) -> String {
        var url: String
        switch endPoint {
        case BOUrlEndPoint.eventPublish:
            url = "\(getBaseServerUrl())/\(BO_SDK_REST_API_EVENTS_PUSH_PATH)"
            break
            
        case BOUrlEndPoint.manifestPull:
            url = "\(getBaseServerUrl())/\(BO_SDK_REST_API_MANIFEST_PULL_PATH)"
            break
        }
        
        let token = BlotoutAnalytics.sharedInstance.token ?? ""
        return "\(url)?token=\(token)"
        
    }
    
    func prepareRequestHeaders() -> [AnyHashable : Any] {
        return [
            BO_ACCEPT: "application/json",
            BO_CONTENT_TYPE: "application/json"
        ]
    }
    
    // This Method check for null value in response data and replace it with empty string, Temp fix for Manifest Response
    func check(forNullValue data: Data?) -> Data? {
        var responseString: String? = nil
        if let data = data {
            responseString = String(data: data, encoding: .utf8)
        }

        if responseString?.contains("null,") ?? false {
            responseString = responseString?.replacingOccurrences(of: "null,", with: "")
        }

        return responseString?.data(using: .utf8)
    }
}
